dali/operators/random/rng_base.h

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_OPERATORS_RANDOM_RNG_BASE_H_
+#define DALI_OPERATORS_RANDOM_RNG_BASE_H_
+
+#include <random>
+#include <vector>
+#include "dali/core/convert.h"
+#include "dali/pipeline/operator/operator.h"
+#include "dali/pipeline/util/batch_rng.h"
+#include "dali/core/static_switch.h"
+#include "dali/operators/util/randomizer.cuh"
+
+namespace dali {
+
+template <typename Backend>
+struct RNGBaseFields;
+
+template <typename Backend, typename Impl>
+class RNGBase : public Operator<Backend> {
+ public:
+  ~RNGBase() override = default;
+
+ protected:
+  explicit RNGBase(const OpSpec &spec)
+      : Operator<Backend>(spec),
+        rng_(spec.GetArgument<int64_t>("seed"), max_batch_size_),
+        backend_specific_(spec.GetArgument<int64_t>("seed"), max_batch_size_) {
+  }
+
+  Impl &This() noexcept { return static_cast<Impl&>(*this); }
+  const Impl &This() const noexcept { return static_cast<const Impl&>(*this); }
+
+  bool CanInferOutputs() const override {
+    return true;
+  }
+
+  bool SetupImpl(std::vector<OutputDesc> &output_desc,
+                 const workspace_t<Backend> &ws) override {
+    if (!spec_.TryGetArgument(dtype_, "dtype"))
+      dtype_ = This().DefaultDataType();
```

Comment:
So if not provided it uses NoType?

Reply:
No, it uses DefaultDataType() which is expected to be implemented by the Implementation.

For normal distribution the default is `float`, but for instance, for coin flip the default should probably be `int`